### PR TITLE
Winding number from libigl in DGtal 

### PR DESCRIPTION
**Pull request by dcoeurjo**
# PR Description

Adding winding number as a model of Euclidean oriented shape.

# Checklist

- [x] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug mode.
- [x] All continuous integration tests pass (Github Actions & appveyor)

### CODE REVIEW

**Comment by dcoeurjo** [comment]
This PR can be reviewed

**Comment by kerautret** [comment]
> This PR can be reviewed

Nice I look it in the coming days-!

**Comment by JacquesOlivierLachaud** [comment]
Looks great ! Quick questions. Since many files have been changed, it is hard to see where to focus the PR. For instance, you are editing a module Winding doc, but I do not see in the generated doc any link to this doc module (in shapes package), and I do not see a new example in the list of examples. Thanks for your help !

**Comment by dcoeurjo** [comment]
The Winding module doc is included in "shapes"
The example is in the polyscope-examples folder

**Comment by dcoeurjo** [comment]
Can someone review this PR? 
I had to edit some doc pages in this PR but the core contribution is in the WindingNumber classes.

**Comment by kerautret** [comment]
@dcoeurjo sorry I miss it, for tomorrow is it fine ? I think I can have some time tomorrow.

**Comment by dcoeurjo** [comment]
👍  thx

File: src/DGtal/shapes/doc/moduleWinding.dox [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+By thresholding the winding function with some values in @f$[0,1]@f$, we can define an iso-surface that approximates the exterior envelope of the shape
+defined by the oriented point cloud. For fast computations, we rely on the [libigl](https://libigl.github.io) of the generalized winding numbers.
+
+
+@warning This class requires to have LIBIGL, CGAL and GMP cmake flags set to true (e.g. `cmake .. -DWITH_LIBIGL!true -DWITH_CGAL=true -DWITH_GMP=true`)
+
+
+\section sectmoduleWinding1 Basic usage
+
+The core of the method is defined in the WindingNumberShape<Space> class. This class is a model of concepts::CEuclideanBoundedShape is the sense that
```
Comment (kerautret) [inline src/DGtal/shapes/doc/moduleWinding.dox:50]: In my generated doc it do appears linked, I don't know if it from me or something wrong ?
Capture d’écran 2024-03-05 à 00 59 44

File: src/DGtal/shapes/doc/moduleWinding.dox [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+@f[
+ w(q) = \sum_i^m a_i \frac{(p_i-q)\cdot n_i}{4\pi\|p_i -q\|^3}
+@f]
+where @f$ a_i@f$ is some area contribution of the point @f$ p_i@f$ to the surface (computed from the projection of the k-nearest neighbors to @f$ p_i@f$ onto the 
+tangent plane at @f$ p_i @f$).
+By thresholding the winding function with some values in @f$[0,1]@f$, we can define an iso-surface that approximates the exterior envelope of the shape
+defined by the oriented point cloud. For fast computations, we rely on the [libigl](https://libigl.github.io) of the generalized winding numbers.
+
+
+@warning This class requires to have LIBIGL, CGAL and GMP cmake flags set to true (e.g. `cmake .. -DWITH_LIBIGL!true -DWITH_CGAL=true -DWITH_GMP=true`)
```
Comment (kerautret) [inline src/DGtal/shapes/doc/moduleWinding.dox:45]:
```suggestion
@warning This class requires to have LIBIGL, CGAL and GMP cmake flags set to true (e.g. `cmake .. -DWITH_LIBIGL=true -DWITH_CGAL=true -DWITH_GMP=true`)
```

File: src/DGtal/shapes/doc/moduleWinding.dox [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+Eigen::MatrixXd &normals = { .... };  //mx3 matrix to encode the normal vector for each point
+
+WindingNumbersShape<Z3i::Space> winding(points, normals);
+
+Z3i::RealPoint q(0.3,0.4,0.0); //a single query point
+
+DGtal::Orientation orientation = winding.orientation(q); //returns DGtal::INSIDE, DGtal::ON or DGtal::OUTSIDE
+@endcode
+
+As the WindingNumberShape class is a model of concepts::CEuclideanBoundedShape, the GaussDigitizer could be used (see @ref moduleShape).
```
Comment (kerautret) [inline src/DGtal/shapes/doc/moduleWinding.dox:69]: Same here,  WindingNumberShape is not linked to the doc page

File: tests/shapes/testWindingNumbersShape.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+/**
+ * @file testlibigl.cpp
```
Comment (kerautret) [inline tests/shapes/testWindingNumbersShape.cpp:18]:
```suggestion
 * @file shapes/testWindingNumbersShape.cpp
```

File: tests/shapes/testWindingNumbersShape.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ * @ingroup Tests
+ * @author David Coeurjolly (\c david.coeurjolly@liris.cnrs.fr )
+ * Laboratoire d'InfoRmatique en Image et Systemes d'information - LIRIS (CNRS, UMR 5205), CNRS, France
+ *
+ * @date 2023/06/14
+ *
+ * Functions for testing class PolygonalCalculus.
```
Comment (kerautret) [inline tests/shapes/testWindingNumbersShape.cpp:25]:
```suggestion
 * Functions for testing class WindingNumbersShape.
```

File: tests/shapes/testWindingNumbersShape.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+#include <DGtal/shapes/GaussDigitizer.h>
+
+///////////////////////////////////////////////////////////////////////////////
+
+using namespace std;
+using namespace DGtal;
+using namespace Z3i;
+
+///////////////////////////////////////////////////////////////////////////////
+// Functions for testing class PolygonalCalculus.
```
Comment (kerautret) [inline tests/shapes/testWindingNumbersShape.cpp:47]:
```suggestion
// Functions for testing class WindingNumbersShape.
```

File: tests/shapes/testlibigl.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <igl/readOBJ.h>
+///////////////////////////////////////////////////////////////////////////////
+
+using namespace std;
+using namespace DGtal;
+using namespace Z3i;
+
+///////////////////////////////////////////////////////////////////////////////
+// Functions for testing class PolygonalCalculus.
```
Comment (kerautret) [inline tests/shapes/testlibigl.cpp:45]:
```suggestion
// Functions for testing libIGL io.
```

**Review by kerautret — COMMENTED** [review]
Nice new features, tested and working directly without any comp issue 👌🏻 (even with CGal add). Just some minor suggestions from generating the doc locally.

File: src/DGtal/shapes/WindingNumbersShape.h [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    /// @param normals a "nx3" matrix for the normal vectors.
+    WindingNumbersShape(ConstAlias<Eigen::MatrixXd> points, 
+                        ConstAlias<Eigen::MatrixXd> normals)
+    {
+      myPoints  = points;
+      myNormals = normals;
+      myPointAreas = Eigen::VectorXd::Ones(myPoints->rows());
+      // Build octree, from libIGL tutorials
+      igl::octree(*myPoints,myO_PI,myO_CH,myO_CN,myO_W);
+      if (points->rows()> 20)
```
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/shapes/WindingNumbersShape.h:93]: done thanks

File: src/DGtal/shapes/doc/moduleWinding.dox (hunk 3 above)
Comment (dcoeurjo) [reply src/DGtal/shapes/doc/moduleWinding.dox:69]: It looks like the page is available (Shape→ ..)  https://dgtal-team.github.io/doc-pr/pr1697/index.html
can you confirm ?

**Comment by dcoeurjo** [comment]
Merging this one, I will check the master for the python issues

**Comment by dcoeurjo** [comment]
thanks @kerautret for the review

**Pull request merged**